### PR TITLE
Adds the ability to catch thrown people.

### PR DESCRIPTION
**Pull request by PowerfulBacon**
## About The Pull Request

Implements the ability to catch thrown people.

## Why It's Good For The Game

Currently thrown carbon mobs deal a completely unblockable stun and force you to drop all your weapons. This adds in a counter, increasing the ceiling for achieving full-tier robustness.

If you have your active hand free and are on throw mode, when a carbon mob hits you from being thrown you will catch it, aggressively grabbing the thrown carbon and disabling throw mode. From there you can throw the carbon back at the aggressor, punishing them if they are unprepared.

## Changelog
:cl:
tweak: Thrown humans can now be caught.
/:cl:

### CODE REVIEW

**Comment by DatBoiTim** [comment]
Someone got robusted after they had a body thrown at them

**Comment by PowerfulBacon** [comment]
Thats not true

**Comment by BrumalLake** [comment]
sure mr. doesn't want firecarry throwing and instead nerfs it anyways

**Comment by PowerfulBacon** [comment]
> 
> 
> sure mr. doesn't want firecarry throwing and instead nerfs it anyways

Well yea if you wan't a nerf, you wouldn't want to buff it.

**Comment by Archanial** [comment]
Ided

**Comment by Sinestia** [comment]
Are throws full power at tier 1 grabs? If not you could upgrade it, I haven't checked. If it requires higher tier grabs then there's frankly enough telegraphing that it shouldn't be a big balance problem. 


Anyways I just want to play catch with a person as the ball so this is 👌🏻

**Comment by PowerfulBacon** [comment]
You can only throw with aggressive grab or higher and is a 2 second uncounterable stun. This adds in the ability to counter the stun by catching the thrown mob, and makes it easy to throw back punishing the thrower.

**Comment by KubeRoot** [comment]
Maybe you should only be able to throw with both hands free? Feels weird to be able to positively yeet a person hard enough to knock somebody down at a distance with just one hand.

Maybe when catching thrown mobs, you should only get the aggressive grab when you have both hands free? Makes sense that you'd be able to brace yourself against the impact with one hand, but maybe not instantly grab them aggressively, and this would also reduce the possibility of playing ping-pong throwing the same guy back and forth until one person misses the grab.

**Comment by qwertyquerty** [comment]
I don't think you should auto have aggressive grab on catch

**Comment by PowerfulBacon** [comment]
> 
> 
> I don't think you should auto have aggressive grab on catch

Throwing in the first place requires aggressive grab, so the victim would have been in an aggressive grab anyway. Aggressive grab allows throwing them, so elevating it to aggressive grab on catch allows you to throw back and punish the person throwing, same way you can catch and throw back normal items

File: code/modules/mob/living/carbon/carbon.dm [hunk 1]
```diff
@@ -69,10 +69,32 @@
...
 		hurt = FALSE
+
+	if(iscarbon(hit_atom) && hit_atom != src)
+		var/mob/living/carbon/victim = hit_atom
+		if(!(victim.movement_type & FLYING))
+			if(victim.can_catch_item())
+				visible_message("<span class='danger'>[victim] catches [src]!</span>",\
+					"<span class='userdanger'>[victim] catches you!</span>")
+				grabbedby(victim, TRUE)
+				victim.setGrabState(GRAB_AGGRESSIVE)
```
Comment (PowerfulBacon) [inline code/modules/mob/living/carbon/carbon.dm]:
```suggestion
```

**Comment by PowerfulBacon** [comment]
Its no longer an aggro grab

**Comment by qwertyquerty** [comment]
What made you change your mind?

File: code/modules/mob/living/carbon/carbon.dm [hunk 2]
```diff
@@ -69,10 +69,31 @@
...
+				log_combat(victim, src, "caught (thrown mob)")
+				return
+			if(hurt)
+				victim.take_bodypart_damage(10,check_armor = TRUE)
+				take_bodypart_damage(10,check_armor = TRUE)
+				victim.Paralyze(20)
+				Paralyze(20)
+				visible_message("<span class='danger'>[src] crashes into [victim], knocking them both over!</span>",\
+					"<span class='userdanger'>You violently crash into [victim]!</span>")
+			playsound(src,'sound/weapons/punch1.ogg',50,1)
```
Comment (qwertyquerty) [inline code/modules/mob/living/carbon/carbon.dm:93]: This sound never plays if the body is caught so catching the body is silent

**Review by qwertyquerty — CHANGES_REQUESTED** [review]

**Comment by Isy232** [comment]
Yes, uncounterable stun is bad.

**Comment by PerishedFraud** [comment]
Uncounterable stun is bad but you have to grab twice in order to throw a person. For the receiver, you can just dodge. 

Catching a person is good, but the winning scenario is dodging or punishing the person for trying to throw someone at you (just shove them to disable grab). As such I suggest that if you catch a person you get slowed for a bit instead of stunned.

File: code/modules/mob/living/carbon/carbon.dm (hunk 2 above)
Comment (PowerfulBacon) [reply code/modules/mob/living/carbon/carbon.dm:93]: Catching the body plays the grab sound effect

**Comment by PowerfulBacon** [comment]
> 
> 
> What made you change your mind?

someone complaining in #general and I could just webedit it.

**Review by qwertyquerty — APPROVED** [review]

**Comment by hotimep** [comment]
grab-fu nerf? nah bitch time to play dodgeball using the clown

**Pull request merged**